core/glrender/src/main/java/edu/cmu/cs/dennisc/render/gl/imp/adapters/GlrBox.java

```diff
@@ -55,266 +55,195 @@
-      break;
-    case 1:
-      origin.x = this.xMax;
-      direction.x = 1;
-      break;
-    case 2:
-      origin.y = this.yMin;
-      direction.y = -1;
-      break;
-    case 3:
-      origin.y = this.yMax;
-      direction.y = 1;
-      break;
-    case 4:
-      origin.z = this.zMin;
-      direction.z = -1;
-      break;
-    case 5:
-      origin.z = this.zMax;
-      direction.z = 1;
-      break;
-    default:
-      rv.setNaN();
-      return rv;
-    }
-    GlrGeometry.getIntersectionInSourceFromPlaneInLocal(rv, ray, m, origin.x, origin.y, origin.z, direction.x, direction.y, direction.z);
-    return rv;
-  }
-
-  @Override
-  protected void propertyChanged(InstanceProperty<?> property) {
-    if (property == owner.xMinimum) {
-      this.xMin = owner.xMinimum.getValue();
-      setIsGeometryChanged(true);
-    } else if (property == owner.xMaximum) {
-      this.xMax = owner.xMaximum.getValue();
-      setIsGeometryChanged(true);
-    } else if (property == owner.yMinimum) {
-      this.yMin = owner.yMinimum.getValue();
-      setIsGeometryChanged(true);
-    } else if (property == owner.yMaximum) {
-      this.yMax = owner.yMaximum.getValue();
-      setIsGeometryChanged(true);
-    } else if (property == owner.zMinimum) {
-      this.zMin = owner.zMinimum.getValue();
-      setIsGeometryChanged(true);
-    } else if (property == owner.zMaximum) {
-      this.zMax = owner.zMaximum.getValue();
-      setIsGeometryChanged(true);
-    } else {
-      super.propertyChanged(property);
-    }
-  }
-
-  private double xMin;
-  private double xMax;
-  private double yMin;
-  private double yMax;
-  private double zMin;
-  private double zMax;
+public class GlrBox extends GlrShape<Box> {
+    private void glBox(Context c, boolean isLightingEnabled, boolean isSubElementRequired, boolean isTexturingEnabled) {
+        //todo: revist vertex ordering
+        int id = 0;
+        double[][] leftFace = {{this.xMin, this.yMin, this.zMax}, {this.xMin, this.yMax, this.zMax}, {this.xMin, this.yMax, this.zMin}, {this.xMin, this.yMin, this.zMin}};
+        double[][] rightFace = {{this.xMax, this.yMin, this.zMin}, {this.xMax, this.yMax, this.zMin}, {this.xMax, this.yMax, this.zMax}, {this.xMax, this.yMin, this.zMax}};
+        double[][] bottomFace = {{this.xMin, this.yMin, this.zMin}, {this.xMax, this.yMin, this.zMin}, {this.xMax, this.yMin, this.zMax}, {this.xMin, this.yMin, this.zMax}};
+        double[][] topFace = {{this.xMin, this.yMax, this.zMax}, {this.xMax, this.yMax, this.zMax}, {this.xMax, this.yMax, this.zMin}, {this.xMin, this.yMax, this.zMin}};
+        double[][] frontFace = {{this.xMin, this.yMax, this.zMin}, {this.xMax, this.yMax, this.zMin}, {this.xMax, this.yMin, this.zMin}, {this.xMin, this.yMin, this.zMin}};
+        double[][] backFace = {{this.xMin, this.yMin, this.zMax}, {this.xMax, this.yMin, this.zMax}, {this.xMax, this.yMax, this.zMax}, {this.xMin, this.yMax, this.zMax}};
+        normal3dMap.put("left", leftNormal3d);
```

Comment:
Filling in the first normal and tex maps can be done once in a static initializer where they are declared below:
```
  …
  private static final Map<String, int[]> normal3dMap = new HashMap<>();
  private static final Map<String, float[][]> texCoordMap = new HashMap<>();
  private static final Map<String, double[][]> vertexMap = new HashMap<>();
  static {
    normal3dMap.put("left", leftNormal3d);
    …
  }
```
Only the vertexMap values are dynamic.